book/src/fork-selection.md

```diff
@@ -0,0 +1,147 @@
+any given time only 1 leader, which can be computed from the ledger itself, can
+propose a fork.  For more details, see [fork generation](fork-generation.md)
+and [leader rotation](leader-rotation.md).
+
+## Algorithm
+
+The basic idea to this approach is to stack consensus votes.  Votes in the
+stack must be for forks that descend from each other and for forks that are
+valid in the ledger they are submitted to.  Each consensus vote has a `lockout`
+in units of time before it can be discarded.  When a vote is added to the
+stack, the lockouts of all the previous votes in the stack are doubled (see
+[Rollback](#Rollback)).  With each new vote, a voter commits the previous votes
+to an ever-increasing lockout.  Since at 32 votes we can consider the system to
+be at `max lockout` any votes with a lockout above `1<<32` are dequeued (FIFO).
+Dequeuing a vote is the trigger for a reward.  If a vote expires before it is
+dequeued, it and all the votes above it are popped (LIFO) from the vote stack.
+The voter needs to start rebuilding the stack from that point.
+
+
+### Rollback
+
+Before a vote is pushed to the stack, all the votes leading up to vote with a
+lower lock time than the new vote are popped.  After rollback lockouts are not
+doubled until the voter catches up to the rollback height of votes.
+
+For example, a vote stack with the following state:
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:| |    4 |         4 |
+2  |                    6 | |    3 |         3 |      4  |                    7
+| |    2 |         2 |      8  |                   10 | |    1 |         1 |
+16 |                   17 |
+
+*Vote 5* is at time 9, and the resulting state is
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:| |    5 |         9 |
+2  |                   11 | |    2 |         2 |      8  |                   10
+| |    1 |         1 |      16 |                   17 |
+
+*Vote 6* is at time 10
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:| |    6 |        10 |
+2 |                   12 | |    5 |         9 |       4 |                   13
+| |    2 |         2 |       8 |                   10 | |    1 |         1 |
+16 |                   17 |
+
+At time 10 the new votes caught up to the previous votes.  But *vote 2* expires
+at 10, so the when *vote 7* at time 11 is applied the votes including and above
+*vote 2* will be popped.
+
+| vote | vote time | lockout | lock expiration time |
+|-----:|----------:|--------:|---------------------:| |    7 |        11 |
+2 |                   13 | |    1 |         1 |      16 |                   17
+|
+
+The lockout for vote 1 will not increase from 16 until the stack contains 5
+votes.
+
+### Slashing and Rewards
+
+The purpose of the lockout is to force a voter to commit opportunity cost to a
```

Comment:
It would have been helpful to read about this purpose much earlier.

Reply:
you want rearrange this around after we finalize the terminology?